zenoh/src/api/session.rs

```diff
@@ -359,142 +367,120 @@ impl Resource {
-            key_expr: TryIntoKeyExpr::try_into(key_expr).map_err(Into::into),
-            #[cfg(feature = "unstable")]
-            reliability: Reliability::DEFAULT,
-            origin: Locality::default(),
-            handler: DefaultHandler::default(),
-        }
-    }
-    fn declare_queryable<'b, TryIntoKeyExpr>(
-        &'s self,
-        key_expr: TryIntoKeyExpr,
-    ) -> QueryableBuilder<'a, 'b, DefaultHandler>
-    where
-        TryIntoKeyExpr: TryInto<KeyExpr<'b>>,
-        <TryIntoKeyExpr as TryInto<KeyExpr<'b>>>::Error: Into<zenoh_result::Error>,
-    {
-        QueryableBuilder {
-            session: self.clone(),
-            key_expr: key_expr.try_into().map_err(Into::into),
-            complete: false,
-            origin: Locality::default(),
-            handler: DefaultHandler::default(),
-        }
-    }
-    fn declare_publisher<'b, TryIntoKeyExpr>(
-        &'s self,
-        key_expr: TryIntoKeyExpr,
-    ) -> PublisherBuilder<'a, 'b>
-    where
-        TryIntoKeyExpr: TryInto<KeyExpr<'b>>,
-        <TryIntoKeyExpr as TryInto<KeyExpr<'b>>>::Error: Into<zenoh_result::Error>,
-    {
-        PublisherBuilder {
-            session: self.clone(),
-            key_expr: key_expr.try_into().map_err(Into::into),
-            encoding: Encoding::default(),
-            congestion_control: CongestionControl::DEFAULT,
-            priority: Priority::DEFAULT,
-            is_express: false,
-            #[cfg(feature = "unstable")]
-            reliability: Reliability::DEFAULT,
-            destination: Locality::default(),
-        }
+impl<'a, T> UndeclarableSealed<&'a Session> for T
+where
+    T: UndeclarableSealed<()>,
+{
+    type Undeclaration = <T as UndeclarableSealed<()>>::Undeclaration;
+
+    fn undeclare_inner(self, _session: &'a Session) -> Self::Undeclaration {
+        self.undeclare_inner(())
     }
-    #[zenoh_macros::unstable]
-    fn liveliness(&'s self) -> Liveliness<'a> {
-        Liveliness {
-            session: self.clone(),
-        }
+}
+
+// NOTE: `UndeclarableInner` is only pub(crate) to hide the `undeclare_inner` method. So we don't
+// care about the `private_bounds` lint in this particular case.
+#[allow(private_bounds)]
+/// A trait implemented by types that can be undeclared.
+pub trait Undeclarable<S = ()>: UndeclarableSealed<S> {}
+
+impl<T, S> Undeclarable<S> for T where T: UndeclarableSealed<S> {}
+
+pub(crate) struct SessionInner {
+    weak_counter: Mutex<usize>,
+    pub(crate) runtime: Runtime,
+    pub(crate) state: RwLock<SessionState>,
+    pub(crate) id: u16,
+    owns_runtime: bool,
+    task_controller: TaskController,
+}
+
+impl fmt::Debug for SessionInner {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Session")
+            .field("id", &self.runtime.zid())
+            .finish()
     }
-    fn info(&'s self) -> SessionInfo<'a> {
-        SessionInfo {
-            session: self.clone(),
-        }
+}
+
+/// A zenoh session.
```

Comment:
I'm just realising now that documentation on `Session` is practically inexistent (this is true also for the published doc: https://docs.rs/zenoh/1.0.0-beta.1/zenoh/struct.Session.html)